src/memex/models/document.py

```diff
@@ -409,6 +405,34 @@ def merge_documents(session, documents, updated=None):
     return master
 
 
+def best_document_uri_from(document_uris):
+    """
+    Return the "best" DocumentURI from the given list of DocumentURIs, or None.
+
+    Return None if there's no http(s) URIs in document_uris.
+
+    """
+    # The first rel-canonical type http(s) URI found.
+    canonical_uri = None
+
+    # The first http(s) URI found that is not a self-claim or rel-canonical.
+    other_uri = None
+
+    for document_uri in document_uris:
+        uri = document_uri.uri
+        type_ = document_uri.type
+
+        if urlparse.urlparse(uri).scheme in ['http', 'https']:
+            if type_ == 'self-claim':
+                return uri
+            elif type_ == 'rel-canonical':
+                canonical_uri = canonical_uri or uri
+            else:
+                other_uri = other_uri or uri
+
+    return canonical_uri or other_uri
```

Comment:
I actually rejected this approach because it loops over the uris (up to) 3 times. But it's not likely to be a long list of uris so that's not likely to matter, so if you think it's clearer then sure I can change it